sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -472,9 +472,12 @@ object TypeCoercion {
         // RHS is the subquery output.
         val rhs = sub.output
 
-        val commonTypes = lhs.zip(rhs).flatMap { case (l, r) =>
-          findCommonTypeForBinaryComparison(l.dataType, r.dataType, conf)
-            .orElse(findTightestCommonType(l.dataType, r.dataType))
+        val commonTypes = lhs.zip(rhs).flatMap {
+          case (l, r) if !l.dataType.isInstanceOf[DecimalType] &&
+            !r.dataType.isInstanceOf[DecimalType] =>
+            findCommonTypeForBinaryComparison(l.dataType, r.dataType, conf)
+              .orElse(findTightestCommonType(l.dataType, r.dataType))
+          case (l, r) => findWiderTypeForDecimal(l.dataType, r.dataType)
```

Comment:
We can clean this part a little bit with the extractor defined in `object DecimalType` for checking expression data types:

```scala
val commonTypes = lhs.zip(rhs).flatMap {
  case (l @ DecimalType(), r @ DecimalType()) =>
    findWiderTypeForDecimal(l.dataType, r.dataType)

  case (l, r) =>
    findCommonTypeForBinaryComparison(l.dataType, r.dataType, conf).orElse {
      findTightestCommonType(l.dataType, r.dataType)
    }
}
```

Reply:
thanks for your suggestion.
To unify the logic of `in` and `inSubquery` mentioned by cloud-fan, I simply call `findWiderTypeForTwo` here.